File: src/debug/entry/javascript/entry.js
var net = require("net");

const fun = require(process.argv[2]);
const testString = process.argv[3];
const paramTypes = process.argv[5].split(",");
const problemNum = parseInt(process.argv[7]);
const debugServerPort = parseInt(process.argv[8]);

const sock = net.connect(
    {
        port: debugServerPort,
        host: "127.0.0.1",
    },
    function() {
        start();
    },
);
sock.setNoDelay(true);

function onClose() {
    setTimeout(() => {
        if (sock != null) {
            sock.end();
        }
        process.exit(0);
    }, 2000);
}

function makeMessage(type, message) {
    return JSON.stringify({
        type,
        message,
        problemNum,
        filePath: process.argv[2],
        testString,
        language: "javascript",
    });
}

function onError(err) {
    sock.write(makeMessage("error", err));
    onClose();
}

function onSuccess() {
    sock.write(makeMessage("success", ""));
    onClose();
}

process.on("uncaughtException", function(err) {
    onError(err.message + "\n" + err.stack);
    throw err;
});

function onParameterError() {
    throw new Error("Parameters parsing error, please check the format of the input parameters");
}

function onParameterTypeError(type) {
    throw new Error(`Unsupported parameter type: ${type}`);
}

function isNumber(num) {
    return num === +num;
}

function isString(str) {
    return typeof str === "string";
}

function isCharacter(str) {
    return isString(str) && str.length === 1;
}

function parseNumber(param) {
    if (!isNumber(param)) {
        onParameterError();
    }
    return param;
}

function parseNumberArray(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }
    param.map(p => {
        if (!isNumber(p)) {
            onParameterError();
        }
    });
    return param;
}

function parseString(param) {
    if (!isString(param)) {
        onParameterError();
    }
    return param;
}

function parseStringArray(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }
    param.map(p => {
        if (!isString(p)) {
            onParameterError();
        }
    });
    return param;
}

function parseStringArrayArray(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }
    param.map(p => {
        if (!parseStringArray(p)) {
            onParameterError();
        }
    });
    return param;
}

function ListNode(val) {
    this.val = val;
    this.next = null;
}

function parseListNode(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }

    let head = null;
    let tail = null;
    param.map(p => {
        const node = new ListNode(p);
        if (head == null) {
            tail = node;
            head = node;
        } else {
            tail.next = node;
            tail = node;
        }
    });
    return head;
}

function parseListNodeArray(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }

    const res = param.map(p => {
        return parseListNode(p);
    });
    return res;
}

function parseNumberArrayArray(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }
    param.map(p => {
        return parseNumberArray(p);
    });
    return param;
}

function parseCharacter(param) {
    if (!isCharacter(param)) {
        onParameterError();
    }
    return param;
}

function parseCharacterArray(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }
    param.map(p => {
        if (!isCharacter(p)) {
            onParameterError();
        }
    });
    return param;
}

function parseCharacterArrayArray(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }
    param.map(p => {
        parseCharacterArray(p);
    });
    return param;
}

function NestedInteger(ni) {
    let nested = [];
    if (Array.isArray(ni)) {
        ni.map(n => {
            nested.push(new NestedInteger(n));
        });
    }

    /**
     * Return true if this NestedInteger holds a single integer, rather than a nested list.
     * @return { boolean }
     */
    this.isInteger = function() {
        if (Array.isArray(ni)) {
            return false;
        }
        return true;
    };

    /**
     * Return the single integer that this NestedInteger holds, if it holds a single integer
     * Return null if this NestedInteger holds a nested list
     * @return { integer }
     */
    this.getInteger = function() {
        if (Array.isArray(ni)) {
            return null;
        }
        return ni;
    };

    /**
     * Return the nested list that this NestedInteger holds, if it holds a nested list
     * Return null if this NestedInteger holds a single integer
     * @return { NestedInteger[] }
     */
    this.getList = function() {
        if (Array.isArray(ni)) {
            return nested;
        }
        return null;
    };
}

function parseNestedIntegerArray(param) {
    return param.map(p => {
        return new NestedInteger(p);
    });
}

function parseMountainArray(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }
    function MountainArray() {
        /**
         * @param {integer} index
         * @return {integer}
         */
        this.get = function(index) {
            return param[index];
        };

        /**
         * @return {integer}
         */
        this.length = function() {
            return param.length;
        };
    }
    return new MountainArray();
}

function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}

function parseTreeNode(param) {
    if (!Array.isArray(param)) {
        onParameterError();
    }

    let root = null;
    const fifo = [];
    let i = 0;
    while (i < param.length) {
        if (i === 0) {
            root = new TreeNode(param[i]);
            i += 1;
            fifo.push(root);
            continue;
        }
        const parent = fifo.shift();
        if (param[i] != null) {
            const left = new TreeNode(param[i]);
            parent.left = left;
            fifo.push(left);
        }
        if (i + 1 < param.length && param[i + 1] != null) {
            const right = new TreeNode(param[i + 1]);
            parent.right = right;
            fifo.push(right);
        }
        i = i + 2;
    }
    return root;
}

function parseParameter(index, type, param) {
    switch (type) {
        case "number":
            return parseNumber(param);
        case "number[]":
            return parseNumberArray(param);
        case "number[][]":
            return parseNumberArrayArray(param);
        case "string":
            return parseString(param);
        case "string[]":
            return parseStringArray(param);
        case "string[][]":
            return parseStringArrayArray(param);
        case "ListNode":
            return parseListNode(param);
        case "ListNode[]":
            return parseListNodeArray(param);
        case "character":
            return parseCharacter(param);
        case "character[]":
            return parseCharacterArray(param);
        case "character[][]":
            return parseCharacterArrayArray(param);
        case "NestedInteger[]":
            return parseNestedIntegerArray(param);
        case "MountainArray":
            return parseMountainArray(param);
        case "TreeNode":
            return parseTreeNode(param);
    }
    const result = parseSpecialParameter(index, type, param);
    if (result == null) {
        onParameterTypeError(type);
    }
    return result;
}

// @@stub-for-code@@

function parseParamsToJson(paramString) {
    let params;
    try {
        const paramsByLine = paramString.split(/\\n/);

        params = paramsByLine.map(str => {
            return JSON.parse(str);
        });

        return params;
    } catch (error) {
        onParameterError();
        return;
    }
}

function start() {
    const paramString = testString.replace(/\\"/g, '"');
    const jsonParams = parseParamsToJson(paramString);

    runUserScript(fun, jsonParams, paramTypes);
    onSuccess();
}

File: src/debug/entry/python3/entry.py
import importlib.util
import sys
from pathlib import Path
import json
import re
import socket

testString = str(sys.argv[2])
funcName = str(sys.argv[3])
paramTypes = str(sys.argv[4]).split(",")
problemNum = int(sys.argv[6])
debugServerPort = int(sys.argv[7])


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def onParameterError():
    raise Exception(
        "Parameters parsing error, please check the format of the input parameters")


def onParameterTypeError(t):
    raise Exception("Unsupported parameter type: {}".format(t))


def isNumber(num):
    return isinstance(num, int)


def isString(s):
    return isinstance(s, str)


def isCharacter(s):
    return isString(s) and len(s) == 1


def isList(li):
    return isinstance(li, list)


def parseNumber(param):
    if (not isNumber(param)):
        onParameterError()
    return param


def parseNumberArray(param):
    if (not isList(param)):
        onParameterError()

    for i in param:
        if (not isNumber(i)):
            onParameterError()
    return param


def parseNumberArrayArray(param):
    if (not isList(param)):
        onParameterError()
    for i in param:
        parseNumberArray(i)
    return param


def parseString(param):
    if (not isString(param)):
        onParameterError()

    return param


def parseStringArray(param):
    if (not isList(param)):
        onParameterError()

    for i in param:
        parseString(i)

    return param


def parseStringArrayArray(param):
    if (not isList(param)):
        onParameterError()

    for i in param:
        parseStringArray(i)

    return param


def parseListNode(param):
    if (not isList(param)):
        onParameterError()

    head = None
    tail = None

    for i in param:
        node = ListNode(i)
        if (head == None):
            tail = node
            head = node
        else:
            tail.next = node
            tail = node

    return head


def parseListNodeArray(param):
    if (not isList(param)):
        onParameterError()

    res = []
    for i in param:
        res.append(parseListNode(i))

    return res


def parseCharacter(param):
    if (not isCharacter(param)):
        onParameterError()

    return param


def parseCharacterArray(param):
    if (not isList(param)):
        onParameterError()

    for i in param:
        parseCharacter(i)

    return param


def parseCharacterArrayArray(param):
    if (not isList(param)):
        onParameterError()

    for i in param:
        parseCharacterArray(i)

    return param


def parseParameter(index, paramType, param):
    switch = {
        "number": lambda x: parseNumber(x),
        "number[]": lambda x: parseNumberArray(x),
        "number[][]": lambda x: parseNumberArrayArray(x),
        "string": lambda x: parseString(x),
        "string[]": lambda x: parseStringArray(x),
        "string[][]": lambda x: parseStringArrayArray(x),
        "ListNode": lambda x: parseListNode(x),
        "ListNode[]": lambda x: parseListNodeArray(x),
        "character": lambda x: parseCharacter(x),
        "character[]": lambda x: parseCharacterArray(x),
        "character[][]": lambda x: parseCharacterArrayArray(x),
    }
    switchfun = switch.get(paramType, 0)

    if switchfun is 0:
        result = parseSpecialParameter(index, paramType, param)
        if result is None:
            onParameterTypeError(paramType)
        return result
    else:
        return switchfun(param)


def loadModule():
    # add module to search path
    parsedPath = Path(sys.argv[1])
    sys.path.append(parsedPath.parent)

    # load module
    spec = importlib.util.spec_from_file_location(parsedPath.stem, sys.argv[1])
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


// @@stub-for-code@@


def start():
    module = loadModule()
    solution = module.Solution()
    func = getattr(solution, funcName)

    lines = testString.split("\\n")

    params = []
    for i, val in enumerate(lines):
        params.append(json.loads(val))

    runUserScript(func, params, paramTypes)


def makeMessage(ty, message):
    return json.dumps({
        "type": ty,
        "message": message,
        "problemNum": problemNum,
        "filePath": sys.argv[1],
        "testString": testString,
        "language": "python",
    }).encode("utf-8")


if __name__ == "__main__":
    sock = socket.socket()
    sock.connect(("127.0.0.1", debugServerPort))
    try:
        start()
        sock.send(makeMessage("success", ""))
    except Exception as identifier:
        sock.send(makeMessage("error", str(identifier)))
        raise identifier
    finally:
        sock.close()
